r1cs-core/src/constraint_system.rs

```diff
@@ -1,235 +1,661 @@
+        }
+        self.num_constraints += 1;
+        let name = self.compute_full_name(name.into());
+        self.constraint_names.push(name);
+        Ok(())
+    }
+
+    /// Return the current namespace
+    pub fn current_namespace(&self) -> &str {
+        &self.current_namespace_path
     }
 
+    /// Enter a new namespace.
     #[inline]
-    fn alloc_input<FN, A, AR>(&mut self, annotation: A, f: FN) -> Result<Variable, SynthesisError>
-    where
-        FN: FnOnce() -> Result<F, SynthesisError>,
-        A: FnOnce() -> AR,
-        AR: Into<String>,
-    {
-        self.0.alloc_input(annotation, f)
+    pub fn enter_namespace(&mut self, name: impl Into<Name>) {
+        self.namespace.push(name.into());
+        self.current_namespace_path = self.namespace.join("/");
     }
 
+    /// Leave a namespace.
     #[inline]
-    fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
-    where
-        A: FnOnce() -> AR,
-        AR: Into<String>,
-        LA: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
-        LB: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
-        LC: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
-    {
-        self.0.enforce(annotation, a, b, c)
+    pub fn leave_namespace(&mut self) {
+        self.namespace.pop();
+        self.current_namespace_path = self.namespace.join("/");
     }
 
-    // Downstream users who use `namespace` will never interact with these
-    // functions and they will never be invoked because the namespace is
-    // never a root constraint system.
+    /// Naively inlines symbolic linear combinations into the linear combinations
+    /// that use them.
+    ///
+    /// Useful for standard pairing-based SNARKs where addition gates are cheap.
+    /// For example, in the SNARKs such as [[Groth16]](https://eprint.iacr.org/2016/260) and
+    /// [[Groth-Maller17]](https://eprint.iacr.org/2017/540), addition gates
+    /// do not contribute to the size of the multi-scalar multiplication, which
+    /// is the dominating cost. (TODO)
+    pub fn inline_all_lcs(&mut self) {
+        let mut inlined_lcs = BTreeMap::new();
+        for (&index, lc) in &self.lc_map {
+            let mut inlined_lc = LinearCombination::new();
+            for &(coeff, var) in lc.iter() {
+                if var.is_lc() {
+                    let lc_index = var.get_lc_index().expect("should be lc");
+                    // If `var` is a `SymbolicLc`, fetch the corresponding
+                    // inlined LC, and substitute it in.
+                    let lc = inlined_lcs.get(&lc_index).expect("should be inlined");
+                    inlined_lc.extend((lc * coeff).0.into_iter());
+                } else {
+                    // Otherwise, it's a concrete variable and so we
+                    // substitute it in directly.
+                    inlined_lc.push((coeff, var));
+                }
+            }
+            inlined_lc.compactify();
```

Comment:
I've seen that this is unimplemented - what is it supposed to do? Make sure that if we have the same variable a few times, we add up its coefficients?

Reply:
Yes, basically after substituting `SymbolicLC`s, we might have duplicate copies of the same variable (with potentially different coefficients), so we call `compactify()` (name open for bike-shedding) to unify these duplicates. I'll add a comment to the code clarifying this.

Reply:
This is now implemented.